LibTest/async/MultiStreamController/onResume_A01_t01.dart

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion void Function()? onResume
+/// The callback which is called when the stream is resumed.
+///
+/// May be set to `null`, in which case no callback will happen.
+///
+/// Pause related callbacks are not supported on broadcast stream controllers.
+///
+/// @description Checks that this callback is called when the stream is resumed.
+/// @author sgrekhov22@gmail.com
+
+import "dart:async";
+import "../../../Utils/expect.dart";
+
+main() {
+  asyncStart(2);
+  var stream = Stream<int>.multi((controller) {
+    Expect.isNull(controller.onResume);
+    controller.onResume = () {
+      controller.close();
+      asyncEnd();
```

Comment:
No need for the `controller.close()` here. It will be delivered after `asyncEnd()` has run, so at least for the second stream, the test is already considered completed.

Reply:
Thank you. Removed.

Reply:
Correct on the behavior of `asyncStart`/`asyncEnd`.
The thing is that the test is considered *done* after the second `asyncEnd`. If nothing has failed before that, the test is *successful*, irrevocably.

Anything that happens after that does not affect the test result, so *preferably* nothing should happen at all. Something throwing after the test is complete is a bug in the test, not a test failure.

In this test, the behavior of `controller.close()` is not what is tested, so there is no need to call it at all.
If we do want to test it, then the `asyncEnd()` should be moved to after that test has been performed.

Reply:
`controller.close()` was removed from this test, so it should be Ok. `asyncEnd()` here checks that `onResume` was really called.

Reply:
> In a way, it does preempt the test result.

I don't think so, the test runner just checks that the subprocess output includes `unittest-suite-success`, but it is certainly still possible for the test to proceed, and to succeed or fail based on later actions.

```dart
import 'dart:async';
import "package:expect/async_helper.dart";

void main() {
  asyncStart();
  asyncEnd();
  throw "Failed!"; // Causes the test to fail, comment it out and it succeeds.
}
```

> Something throwing after the test is complete is a bug in the test, not a test failure

It isn't quite fair to say that this is what is happening in examples like the one above, or in LibTest/async/MultiStreamController/onResume_A01_t01.dart.

The test will succeed or fail according to the test runner (which is what counts), and there's no difference between failures happening before or after the final invocation of `asyncEnd()`.

What `asyncStart(_)`/`asyncEnd()` will do is simply to support behavior that is detected as a test run failure in the case where the stand/end invocation counts differ. This can be used to detect, in particular, that a given piece of code which was expected to run did not actually run.

> In this test, the behavior of `controller.close()` is not what is tested, so there is no need to call it at all.

OK, that's good to know! Perhaps `controller.close()` should then be deleted from all tests which aren't testing this, and then we should make sure there is one or more tests targeting the effects of invoking `controller.close()`.

> If we do want to test it, then the `asyncEnd()` should be moved to after that test has been performed.

That makes sense, too. It shouldn't matter where it occurs in a piece of code that does not include scheduling points (like `await`), but it's safer to put it at the end, indicating that _all_ of the given code has been executed when the test terminates.

Reply:
Closing this thread. I don't think there's anything substantial to resolve here. ;-)

Reply:
The thread is closed but I can't find my comment here. Adding just in case.

There was a long discussion about it. See
https://github.com/dart-lang/co19/pull/2400#issuecomment-1831769630